ledger/src/blockstore.rs

```diff
@@ -3815,21 +3854,34 @@ where
     Ok(())
 }
 
+// Update the meta for a slot whose parent is known to be connected.
+// Returns a bool indicating whether this slot is now connected which would
+// indicate that it's children are also in need of update.
+fn update_child_from_connected_parent(meta: &mut SlotMeta) -> bool {
+    meta.set_parent_connected(true);
```

Comment:
If this function is only for updating purposes, we should first check `meta.is_connected`.
If `meta.is_connected` is true, then we simply return false, indicating there is no need to update.

Reply:
So I think this is kind of similar to your notes about the assertions for sanity checks.

This function gets called on slot `S` when `S`'s parent `P` has been marked connected. By definition, if `P` just became connected, then `S` could not have (or at least should not have) already been connected.

I guess one hypothetical scenario where my above statement would break down is if:
- User loads from snapshot at slot `S1` and progresses onwards
- User stops, and restart from a snapshot at slot `S2` where `S2 < S1`
   - `S2 < S1` should sound weird, validators don't delete their later snapshots to start at an older one
- User now replays from `S2` onward, where slots > `S2` will get marked connected
- Eventually, replay progresses from the second start hits `S1` again. `S1` is already connected but with existing logic, we would mark it connected again

There is nothing preventing someone from doing what I just described, but I don't why any operator running a node regularly would do something like this

Reply:
Still don't think this scenario should occur often, but made the code defensive to avoid rewriting the same bit